File: LeetCode/1345. Jump Game IV/solution.py
class Solution:
    # def minJumps(self, arr: List[int]) -> int:
    #     g = defaultdict(list)
    #     for i, a in enumerate(arr):
    #         g[a].append(i)
    #     visited = set([0])
    #     q = deque([0])
    #     step = 0
    #     while q:
    #         n = len(q)
    #         for _ in range(len(q)):
    #             cur = q.popleft()
    #             if cur == len(arr)-1:
    #                 return step
    #             for nxt in [-1,1]:
    #                 nxt += cur
    #                 if 0 <= nxt < len(arr) and nxt not in visited:
    #                     visited.add(nxt)
    #                     q.append(nxt)
    #             while g[arr[cur]]:
    #                 nxt = g[arr[cur]].pop()
    #                 if nxt not in visited:
    #                     visited.add(nxt)
    #                     q.append(nxt)
    #         step += 1
    #     return -1
    
    def minJumps(self, A):
        indices = collections.defaultdict(list)
        for i, a in enumerate(A):
            indices[a].append(i)
        done, now = {-1}, {0}
        for steps in itertools.count():
            done |= now
            if len(A) - 1 in done:
                return steps
            now = {j for i in now for j in [i-1, i+1] + indices.pop(A[i], [])} - done

File: LeetCode/1345. Jump Game IV/solution.cpp
class Solution {
public:
    int minJumps(vector<int>& arr) {
        int n = arr.size();
        vector<int> dp(n,n);
        unordered_map<int,vector<int>> mp;
        for (int i = 0; i < n; i++)
            mp[arr[i]].push_back(i);
        dp[0] = 0;
        queue<int> q;
        q.push(0);
        while (!q.empty()){
            int i = q.front(); q.pop();
            if (i == n-1)
                return dp[i];
            if (i-1 >= 0 && dp[i-1] > dp[i]+1){
                dp[i-1] = dp[i]+1;
                q.push(i-1);
            }
            if (i+1 < n && dp[i+1] > dp[i]+1){
                dp[i+1] = dp[i]+1;
                q.push(i+1);
            }
            for (auto j : mp[arr[i]]){
                if (j == i)
                    continue;
                if (dp[j] > dp[i]+1){
                    dp[j] = dp[i]+1;
                    q.push(j);
                }
            }
            mp[arr[i]].clear();
        }
        return -1;
    }
};